File: 2017/day12/day12.py
from typing import Optional, Set, Dict
from collections import defaultdict


class Plumber:
    def getInput(self) -> Dict[int, Set[int]]:
        inputFile = "input-test.txt" if self.useTest else "input.txt"
        pipeMap = defaultdict(set)
        with open(inputFile, "r") as file1:
            for line in file1:
                line = line.strip().split(" <-> ")
                for end in line[1].split(","):
                    pipeMap[int(line[0])].add(int(end))
        return pipeMap

    def __init__(self, useTest: Optional[bool] = False) -> None:
        self.useTest = useTest
        self.pipeMap = self.getInput()
        self.count = self.groupCount = 0

    def countID(self, pipeID: int, visited: Set[int]) -> None:
        if pipeID in visited:
            return
        visited.add(pipeID)
        self.count += 1
        for pipe in self.pipeMap[pipeID]:
            self.countID(pipe, visited)

    def getCount(self) -> int:
        self.count = 0
        self.countID(0, set())
        return self.count

    def getGroups(self) -> int:
        visited = set()
        for pipeID in self.pipeMap.keys():
            if pipeID not in visited:
                self.groupCount += 1
                self.countID(pipeID, visited)
        return self.groupCount


if __name__ == "__main__":
    plumber = Plumber(True)
    print("Day 12 part 1:", plumber.getCount())
    print("Day 12 part 2:", plumber.getGroups())

File: 2025/day12/day12.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Christmas struct {
	useTest  bool
	part1    int
	part2    int
	presents []Present
	grids    []Grid
}

type Present struct {
	coords map[Coord]bool
	size   int
}

type Coord struct {
	x int
	y int
}

type Grid struct {
	width  int
	height int
	sizes  []int
}

func (this *Christmas) getInput() {
	inputFile := "input.txt"
	if this.useTest {
		inputFile = "input-test.txt"
	}

	file, _ := os.Open(inputFile)
	scanner := bufio.NewScanner(file)
	currCoords := []string{}

	numShapes := 0
	for scanner.Scan() {
		line := scanner.Text()
		if line == "" {

			if numShapes == 6 {
				numShapes += 1
			}
			continue
		}
		if numShapes > 6 {

			parts := strings.Split(line, ": ")
			dimensions := strings.Split(parts[0], "x")
			x, _ := strconv.Atoi(dimensions[0])
			y, _ := strconv.Atoi(dimensions[1])

			sizeStr := strings.Split(parts[1], " ")
			sizes := []int{}
			for _, size := range sizeStr {
				s, _ := strconv.Atoi(string(size))
				sizes = append(sizes, s)

			}
			this.grids = append(this.grids, Grid{x, y, sizes})

		} else {

			if line[1] == ':' {
				numShapes += 1
				present := parseCoord(currCoords)
				this.presents = append(this.presents, Present{present, len(present)})
				currCoords = []string{}

			} else {
				currCoords = append(currCoords, line)
			}
		}
	}
	defer file.Close()
}

func parseCoord(grid []string) map[Coord]bool {
	coords := make(map[Coord]bool)
	for y, row := range grid {
		for x, val := range row {
			if val == '#' {
				coords[Coord{x, y}] = true
			}
		}
	}
	return coords
}

// A bit lame solution using information heuristic (i.e approximation)
func (this *Christmas) getPart1() int {
	count := 0
	for _, g := range this.grids {
		allowable := (g.height / 3 * g.width / 3)

		total := 0
		for _, amount := range g.sizes {
			total += amount
		}

		// Approximate
		if total > allowable {
			continue
		}
		count += 1
	}
	return count
}

func main() {
	useTest := flag.Bool("test", false, "use test input file")
	flag.Parse()
	christmas := &Christmas{
		useTest: *useTest,
	}
	christmas.getInput()
	fmt.Println("Day 12 part 1:", christmas.getPart1())
}
